notifier/notifier.go

```diff
@@ -264,6 +264,14 @@ func (n *Manager) ApplyConfig(conf *config.Config) error {
 			return err
 		}
 
+		// If alertmanagers already exists, don't skip it, in reload step will be refreshed new targets.
+		oldAmSets, ok := n.alertmanagers[k]
```

Comment:
I don’t think we should rely on `k` for this, given how it’s constructed: https://github.com/prometheus/prometheus/blob/5c417684f87f2e7f55b172b3e5e5c96ac9bc2e2c/config/config.go#L910-L917. It only takes the config position in the array into account. Reordering the configs or changing the entire SD config for an entry may temporarily mess with its `ams` which isn't desired.

I'd expect a test case for that as well.

I think we'll need to compare the configs as we do in other places. If the `AlertmanagerConfig` is already known, temporarily use the corresponding `alertmanagerSet`'s `ams`.

We'll need to agree on certain behaviors:
- Should we set the existing `droppedAms` as well?
- Perhaps we don't want the entire `AlertmanagerConfig` to be the same. Maybe equal `ServiceDiscoveryConfigs` (and `RelabelConfigs`) are sufficient, but it wouldn't be safe.